File: WebClient SDK/src/default/tiles/testProcessTileExtension.ts
import { TileExtension, ITileGlobalExtensionContext, Tile, TileGroups, TileEvaluationEventArgs } from 'tessa/ui/tiles';
import { ICardModel, CardSavingRequest, CardSavingMode } from 'tessa/ui/cards';
import { hasFlag, createTypedField, DotNetType } from 'tessa/platform';
import { CardTypeFlags } from 'tessa/cards/types';
import { CardStoreMode } from 'tessa/cards';
import { UIContext, tryGetFromInfo } from 'tessa/ui';
import { IStorage } from 'tessa/platform/storage';
import { WorkflowQueue } from 'tessa/cards/workflow';

export class TestProcessTileExtension extends TileExtension {

  public initializingGlobal(context: ITileGlobalExtensionContext) {
    const contextSource = context.workspace.leftPanel.contextSource;
    context.workspace.leftPanel.tiles.push(
      new Tile({
        name: 'StartTestProcess',
        caption: '$KrTest_TestApprovalTile',
        icon: 'ta icon-thin-127',
        contextSource,
        command: TestProcessTileExtension.startTestProcessAction,
        group: TileGroups.Cards,
        order: 6,
        evaluating: TestProcessTileExtension.enableOnTestTypesAndNoProcesses
      }),
      new Tile({
        name: 'SendTestSignal',
        caption: '$KrTest_TestSignalTile',
        icon: 'ta icon-thin-229',
        contextSource,
        command: TestProcessTileExtension.sendTestSignalAction,
        group: TileGroups.Cards,
        order: 6,
        evaluating: TestProcessTileExtension.enableOnTestTypesAndHasProcesses
      })
    );
  }

  private static enableOnTestTypesAndNoProcesses(e: TileEvaluationEventArgs) {
    const editor = e.currentTile.context.cardEditor;
    let model: ICardModel;
    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor
      && !!(model = editor.cardModel!)
      && hasFlag(model.cardType.flags, CardTypeFlags.AllowTasks)
      && model.cardType.name === 'Car'
      && model.card.storeMode === CardStoreMode.Update
      && model.card.sections.get('WorkflowProcesses')!.rows.length === 0
    );
  }

  private static enableOnTestTypesAndHasProcesses(e: TileEvaluationEventArgs) {
    const editor = e.currentTile.context.cardEditor;
    let model: ICardModel;
    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor
      && !!(model = editor.cardModel!)
      && hasFlag(model.cardType.flags, CardTypeFlags.AllowTasks)
      && model.cardType.name === 'Car'
      && model.card.storeMode === CardStoreMode.Update
      && model.card.sections.get('WorkflowProcesses')!.rows.length > 0
    );
  }

  private static async startTestProcessAction() {
    const context = UIContext.current;
    const editor = context.cardEditor;

    if (!editor || !editor.cardModel) {
      return;
    }

    await editor.saveCard(
      context,
      {
        '.startProcess': createTypedField('TestProcess', DotNetType.String)
      }
    );
  }

  private static async sendTestSignalAction() {
    const context = UIContext.current;
    const editor = context.cardEditor;

    if (!editor || !editor.cardModel) {
      return;
    }

    await editor.saveCard(
      context,
      undefined,
      new CardSavingRequest(
        CardSavingMode.RefreshOnSuccess,
        (card) => {
          const info = card.info;
          let storage: IStorage;
          let queue: WorkflowQueue;
          if ((storage = tryGetFromInfo(info, '.workflowQueue', null)!)) {
            queue = new WorkflowQueue(storage);
          } else {
            queue = new WorkflowQueue(storage);
            info['.workflowQueue'] = queue.getStorage();
          }

          queue.addSignal({processTypeName: 'Main', name: 'TestSignal'});
        }
      )
    );
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/TestProcessTileExtension.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Cards.Workflow;
using Tessa.Extensions.Default.Shared.Workflow.TestProcess;
using Tessa.Platform.Collections;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;

namespace Tessa.Extensions.Default.Client.Tiles
{
    /// <summary>
    /// Расширения для бизнес-процесса TestProcess.
    /// </summary>
    public sealed class TestProcessTileExtension :
        TileExtension
    {
        #region Evaluating Handlers

        private static void EnableOnTestTypesAndNoProcesses(object sender, TileEvaluationEventArgs e)
        {
            ICardEditorModel editor = e.CurrentTile.Context.CardEditor;
            ICardModel model;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                editor != null
                && (model = editor.CardModel) != null
                && model.CardType.Flags.Has(CardTypeFlags.AllowTasks)
                && model.CardType.Name == "Car"
                && model.Card.StoreMode == CardStoreMode.Update
                && model.Card.Sections["WorkflowProcesses"].Rows.Count == 0);
        }


        private static void EnableOnTestTypesAndHasProcesses(object sender, TileEvaluationEventArgs e)
        {
            ICardEditorModel editor = e.CurrentTile.Context.CardEditor;
            ICardModel model;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                editor != null
                && (model = editor.CardModel) != null
                && model.CardType.Flags.Has(CardTypeFlags.AllowTasks)
                && model.CardType.Name == "Car"
                && model.Card.StoreMode == CardStoreMode.Update
                && model.Card.Sections["WorkflowProcesses"].Rows.Count > 0);
        }

        #endregion

        #region Command Actions

        private static async void StartTestProcessActionAsync(object parameters)
        {
            IUIContext context = UIContext.Current;
            ICardEditorModel editor = context.CardEditor;

            if (editor == null || editor.CardModel == null)
            {
                return;
            }

            await editor.SaveCardAsync(
                context,
                new Dictionary<string, object>()
                    .SetStartingProcessName(TestProcessHelper.ProcessName));
        }


        private static async void SendTestSignalActionAsync(object parameters)
        {
            IUIContext context = UIContext.Current;
            ICardEditorModel editor = context.CardEditor;

            if (editor == null || editor.CardModel == null)
            {
                return;
            }

            await editor.SaveCardAsync(
                context,
                request: new CardSavingRequest(cardModifierActionAsync:
                    async (card, ct) =>
                    {
                        WorkflowQueue queue = card.GetWorkflowQueue();

                        queue.AddSignal(
                            TestProcessHelper.MainSubProcess,
                            TestProcessHelper.TestSignal);
                    }));
        }

        #endregion

        #region Base Overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            ITileContextSource contextSource = context.Workspace.LeftPanel;

            context.Workspace.LeftPanel.Tiles.AddRange(
                new Tile(
                    "StartTestProcess",
                    TileHelper.SplitCaption("$KrTest_TestApprovalTile"),
                    context.Icons.Get("Thin127"),
                    contextSource,
                    new DelegateCommand(StartTestProcessActionAsync),
                    TileGroups.Cards,
                    order: 6,
                    evaluating: EnableOnTestTypesAndNoProcesses),

                new Tile(
                    "SendTestSignal",
                    TileHelper.SplitCaption("$KrTest_TestSignalTile"),
                    context.Icons.Get("Thin229"),
                    contextSource,
                    new DelegateCommand(SendTestSignalActionAsync),
                    TileGroups.Cards,
                    order: 6,
                    evaluating: EnableOnTestTypesAndHasProcesses))
                ;

            return Task.CompletedTask;
        }

        #endregion
    }
}
